app/src/main/cpp/HandModels.cpp

```diff
@@ -0,0 +1,356 @@
+
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboJointWeights));
+  VRB_GL_CHECK(glEnableVertexAttribArray((GLuint)m.aJointWeights));
+  VRB_GL_CHECK(glVertexAttribPointer((GLuint)m.aJointWeights, 4, GL_FLOAT, GL_FALSE, 0, nullptr));
+
+  VRB_GL_CHECK(glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, state.iboIndices));
+  VRB_GL_CHECK(glDrawElements(GL_TRIANGLES, state.indexCount, GL_UNSIGNED_SHORT, nullptr));
+
+  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+  if (!enabled)
+    VRB_GL_CHECK(glDisable(GL_DEPTH_TEST));
+}
+
+void HandModels::UpdateHandModel(const Controller& aController) {
+  VRB_LOG("HandModel: %u vertices, %u indices", aController.handMesh.vertexCount, aController.handMesh.indexCount);
+
+  if (aController.index >= m.handGLState.size())
+    m.handGLState.resize(aController.index + 1);
+
+  const ControllerDelegate::HandMesh& mesh = aController.handMesh;
+
+  HandGLState& state = m.handGLState.at(aController.index);
+  state.jointCount = mesh.jointCount;
+  state.vertexCount = mesh.vertexCount;
+  state.indexCount = mesh.indexCount;
+
+  if (state.iboIndices == 0) {
+    VRB_GL_CHECK(glGenBuffers(1, &state.vboPosition));
+    VRB_GL_CHECK(glGenBuffers(1, &state.vboNormal));
+    VRB_GL_CHECK(glGenBuffers(1, &state.vboJointIndices));
+    VRB_GL_CHECK(glGenBuffers(1, &state.vboJointWeights));
+
+    VRB_GL_CHECK(glGenBuffers(1, &state.iboIndices));
+  }
+
+  // Positions VBO
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboPosition));
+  VRB_GL_CHECK(glBufferData(GL_ARRAY_BUFFER, state.vertexCount * 3 * sizeof(float), mesh.positions.data(), GL_STATIC_DRAW));
+
+  // Normals VBO
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboNormal));
+  VRB_GL_CHECK(glBufferData(GL_ARRAY_BUFFER, state.vertexCount * 3 * sizeof(float), mesh.normals.data(), GL_STATIC_DRAW));
+
+  // Joint indices VBO
+  std::vector<float> jointIndices;
+  jointIndices.resize(state.vertexCount * 4);
+  for (int i = 0; i < state.vertexCount; i++) {
+      jointIndices[i * 4 + 0] = mesh.jointIndices[i].x;
+      jointIndices[i * 4 + 1] = mesh.jointIndices[i].y;
+      jointIndices[i * 4 + 2] = mesh.jointIndices[i].z;
+      jointIndices[i * 4 + 3] = mesh.jointIndices[i].w;
+  }
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboJointIndices));
+  VRB_GL_CHECK(glBufferData(GL_ARRAY_BUFFER, state.vertexCount * 4 * sizeof(float), jointIndices.data(), GL_STATIC_DRAW));
+
+  // Joint weights VBO
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboJointWeights));
+  VRB_GL_CHECK(glBufferData(GL_ARRAY_BUFFER, state.vertexCount * 4 * sizeof(float), mesh.jointWeights.data(), GL_STATIC_DRAW));
+
+  // Indices IBO
+  VRB_GL_CHECK(glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, state.iboIndices));
+  VRB_GL_CHECK(glBufferData(GL_ELEMENT_ARRAY_BUFFER, state.indexCount * 1 * sizeof(uint16_t), mesh.indices.data(), GL_STATIC_DRAW));
+
+  // Joint bind matrices
+  state.bindMatrices.resize(XR_EXT_HAND_TRACKING_NUM_JOINTS);
```

Comment:
Hmm that variable is from OpenXR, we should not use it here directly dince this code is in theory generic.

Reply:
It is not from OpenXR, I declared it at the top of HandModels.cpp. We need to have a maximum allocation for the joints arrays.